File: dsa/concepts/10_trees/bst/python/l653__two_sum_iv_input_is_a_bst.py
# Tree / HashSet
# Problem: https://leetcode.com/problems/two-sum-iv-input-is-a-bst/
# Idea:
# 1. Similar to two-sum problem.
# 2. Use a set to store visited node values.
# 3. Traverse BST (preorder) and check if k - node.val exists in set.
# 4. Alternative: Use BST iterators with two-pointer technique on sorted nodes.
# Time: O(n)
# Space: O(n) for set

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

def findTarget(root, k):
    seen = set()
    
    def dfs(node):
        if not node:
            return False
        if (k - node.val) in seen:
            return True
        seen.add(node.val)
        return dfs(node.left) or dfs(node.right)
    
    return dfs(root)


# -------- Test --------
# Helper function to insert into BST
def insertIntoBST(root, val):
    if not root:
        return TreeNode(val)
    if val < root.val:
        root.left = insertIntoBST(root.left, val)
    else:
        root.right = insertIntoBST(root.right, val)
    return root

root = None
for val in [10, 2, 4, 6, 5, 15, 3]:
    root = insertIntoBST(root, val)

print(findTarget(root, 10))  # Expected: True (e.g., 4 + 6 = 10)

File: dsa/topics/10_trees/bst/java/L653_TwoSumIvInputIsABst.java
import java.util.HashSet;

import com.algos.topics.models.TreeNode;

public class L653_TwoSumIvInputIsABst {

	public static void main(String[] args) {
		L653_TwoSumIvInputIsABst obj = new L653_TwoSumIvInputIsABst();

		L701_InsertIntoABinarySearchTree bst = new L701_InsertIntoABinarySearchTree();
		TreeNode root = bst.insertIntoBST(null, 10);
		bst.insertIntoBST(root, 2);
		bst.insertIntoBST(root, 4);
		bst.insertIntoBST(root, 6);
		bst.insertIntoBST(root, 5);
		bst.insertIntoBST(root, 15);
		bst.insertIntoBST(root, 3);

		System.out.println(obj.findTarget(root, 10));
	}

	// Problem: https://leetcode.com/problems/two-sum-iv-input-is-a-bst/
	// Idea: Similar like two sum. Use HashSet and pass it to each recursive call.
	// Do preorder traversal in BST and check if k-root.val is exist in set.
	//
	// Alternative: You can implement the same using BST iterator by implementing
	// next and before.
	// Next and before will be traversing on sorted nums, use two pointed logic to
	// find matching nums.
	// Time: O(n)
	// space: O(1)
	public boolean findTarget(TreeNode root, int k) {
		HashSet<Integer> set = new HashSet<>();
		return findTarget(root, k, set);
	}

	private boolean findTarget(TreeNode root, int k, HashSet<Integer> set) {
		if (root == null)
			return false;

		if (set.contains(k - root.val))
			return true;

		set.add(root.val);
		return findTarget(root.left, k, set) || findTarget(root.right, k, set);
	}

}
